disttask/framework/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,412 @@
+				if err != nil {
+					logutil.BgLogger().Warn("get new task failed", zap.Error(err))
+					break
+				}
+				// There are currently no global tasks to work on.
+				if gTask == nil {
+					break
+				}
+				if d.isRunningGlobalTask(gTask.ID) {
+					break
+				}
+				if gTask.State == proto.TaskStateReverting {
+					d.setRunningGlobalTasks(gTask)
+					cnt++
+					continue
+				}
+
+				err = d.processNormalFlow(gTask, true)
+				if err != nil {
+					d.delRunningGlobalTasks(gTask.ID)
+				}
+				d.setRunningGlobalTasks(gTask)
+				cnt++
+			}
+		}
+	}
+}
+
+func (d *dispatcher) detectTask(gTask *proto.Task) (isFinished bool, subTaskErr string) {
+	// TODO: Consider putting the following operations into a transaction.
+	// TODO: Consider collect some information about the tasks.
+	if gTask.State != proto.TaskStateReverting {
+		cnt, err := d.subTaskMgr.GetSubtaskInStatesCnt(gTask.ID, proto.TaskStateFailed)
+		if err != nil {
+			logutil.BgLogger().Warn("check task failed", zap.Int64("task ID", gTask.ID), zap.Error(err))
+			return false, ""
+		}
+		if cnt > 0 {
+			return false, proto.TaskStateFailed
+		}
+
+		cnt, err = d.subTaskMgr.GetSubtaskInStatesCnt(gTask.ID, proto.TaskStatePending, proto.TaskStateRunning)
+		if err != nil {
+			logutil.BgLogger().Warn("check task failed", zap.Int64("task ID", gTask.ID), zap.Error(err))
+			return false, ""
+		}
+		if cnt > 0 {
+			logutil.BgLogger().Info("check task, subtasks aren't finished", zap.Int64("task ID", gTask.ID), zap.Int64("cnt", cnt))
+			return false, ""
+		}
+		return true, ""
+	}
+
+	cnt, err := d.subTaskMgr.GetSubtaskInStatesCnt(gTask.ID, proto.TaskStateRevertPending, proto.TaskStateReverting)
+	if err != nil {
+		logutil.BgLogger().Warn("check task failed", zap.Int64("task ID", gTask.ID), zap.Error(err))
+		return false, ""
+	}
+	if cnt > 0 {
+		return false, ""
+	}
+	return true, ""
+}
+
+// DetectTaskLoop monitors the status of the subtasks.
+func (d *dispatcher) DetectTaskLoop() {
+	logutil.BgLogger().Info("detection task loop start")
+	ticker := time.NewTicker(checkTaskFinishedInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-d.ctx.Done():
+			logutil.BgLogger().Info("detection task loop exits", zap.Error(d.ctx.Err()))
+			return
+		case <-ticker.C:
+			gTasks := d.getRunningGlobalTasks()
+			// TODO: Consider asynchronous processing.
+			for _, gTask := range gTasks {
+				stepIsFinished, errStr := d.detectTask(gTask)
+				// The global task isn't finished and failed.
+				if !stepIsFinished && errStr == "" {
+					logutil.BgLogger().Debug("detection, this task keeps current state",
+						zap.Int64("taskID", gTask.ID), zap.String("state", gTask.State))
+					continue
+				}
+
+				d.processFlow(gTask, errStr)
+			}
+		}
+	}
+}
+
+func (d *dispatcher) processFlow(gTask *proto.Task, errStr string) {
+	var err error
+	if errStr != "" {
+		// Found an error when task is running.
+		logutil.BgLogger().Info("detection, handle an error", zap.Int64("taskID", gTask.ID))
```

Comment:
When `errStr != "" && gTask.State == proto.TaskStateReverting` , it seems the dispatcher will keep adding new revert tasks. Is it by design?

Reply:
Yes.